### PR TITLE
New Resource: azurerm_monitor_metric_alert

### PR DESCRIPTION
**Pull request by JunyiYi**
#### Description

In this pull request, I implemented a new Azure monitoring resource in M4 called metric alerts. It will monitor the metrics (such as transactions count, CPU percentage, etc.) of one or more resources. Together with the [`azurerm_monitor_activitylog_alert`](https://github.com/terraform-providers/terraform-provider-azurerm/pull/1989), they are called the [unified alerts](https://docs.microsoft.com/en-us/azure/monitoring-and-diagnostics/monitoring-overview-unified-alerts) in Azure monitor.

#### Related Issues

This resource will realize the requirements specified in #2007  .

#### References

Resource Description: https://docs.microsoft.com/en-us/azure/monitoring-and-diagnostics/alert-metric-overview?toc=/azure/azure-monitor/toc.json
REST API Reference: https://docs.microsoft.com/en-us/rest/api/monitor/metricalerts/createorupdate

#### Remarks

This resource is different than [`azurerm_metric_alertrule`](https://www.terraform.io/docs/providers/azurerm/r/metric_alertrule.html) (which is called [classic alerts](https://docs.microsoft.com/en-us/azure/monitoring-and-diagnostics/monitoring-overview-alerts-classic?toc=/azure/azure-monitor/toc.json) in Azure). Actually this resource deprecates the old one.

### CODE REVIEW

File: azurerm/resource_arm_monitor_metric_alert.go [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+		},
+	}
+}
+
+func resourceArmMonitorMetricAlertCreateOrUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).monitorMetricAlertsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	name := d.Get("name").(string)
+	resGroup := d.Get("resource_group_name").(string)
```
Comment (katbyte) [inline azurerm/resource_arm_monitor_metric_alert.go]: Could we use `resourceGroup` rather then `resGroup` here?

File: azurerm/resource_arm_monitor_metric_alert.go [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+
+func resourceArmMonitorMetricAlertRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).monitorMetricAlertsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resGroup := id.ResourceGroup
```
Comment (katbyte) [inline azurerm/resource_arm_monitor_metric_alert.go]: Could we use `resourceGroup` rather then `resGroup` here?

File: azurerm/resource_arm_monitor_metric_alert.go [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+
+func resourceArmMonitorMetricAlertDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).monitorMetricAlertsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resGroup := id.ResourceGroup
```
Comment (katbyte) [inline azurerm/resource_arm_monitor_metric_alert.go]: same here

File: azurerm/resource_arm_monitor_metric_alert.go [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+	if resp, err := client.Delete(ctx, resGroup, name); err != nil {
+		if !response.WasNotFound(resp.Response) {
+			return fmt.Errorf("Error deleting metric alert %q (resource group %q): %+v", name, resGroup, err)
+		}
+	}
+
+	return nil
+}
+
+func expandMonitorMetricAlertStringArray(v []interface{}) *[]string {
```
Comment (katbyte) [inline azurerm/resource_arm_monitor_metric_alert.go]: This is a pretty common piece of code, might be work moving into utils?

File: azurerm/resource_arm_monitor_metric_alert.go [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+		AllOf:     &criterias,
+		OdataType: insights.OdataTypeMicrosoftAzureMonitorSingleResourceMultipleMetricCriteria,
+	}
+}
+
+func expandMonitorMetricAlertAction(v []interface{}) *[]insights.MetricAlertAction {
+	actions := make([]insights.MetricAlertAction, 0)
+	for _, actionValue := range v {
+		val := actionValue.(map[string]interface{})
+
```
Comment (katbyte) [inline azurerm/resource_arm_monitor_metric_alert.go]: might i suggest the following whitespace:

```go
val := actionValue.(map[string]interface{})
agID := val["action_group_id"].(string)

props := make(map[string]*string)
if propsValue, ok := val["webhook_properties"]; ok {
```

File: azurerm/resource_arm_monitor_metric_alert.go [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+		actions = append(actions, insights.MetricAlertAction{
+			ActionGroupID:     utils.String(agID),
+			WebhookProperties: props,
+		})
+	}
+	return &actions
+}
+
+func flattenMonitorMetricAlertStringArray(v *[]string) []interface{} {
```
Comment (katbyte) [inline azurerm/resource_arm_monitor_metric_alert.go]: Same comment as above. Might be useful to have this in `utils`

File: azurerm/resource_arm_monitor_metric_alert.go [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+		for _, item := range *v {
+			result = append(result, item)
+		}
+	}
+	return result
+}
+
+func flattenMonitorMetricAlertCriteria(v insights.BasicMetricAlertCriteria) []interface{} {
+	result := make([]interface{}, 0)
+	if v != nil {
```
Comment (katbyte) [inline azurerm/resource_arm_monitor_metric_alert.go]: how about we invert this?

```go
if v == nil {
    return result
}
```

File: azurerm/resource_arm_monitor_metric_alert.go [hunk 8]
```diff
@@ -0,0 +1,450 @@
...
+		if val, ok := v.AsMetricAlertSingleResourceMultipleMetricCriteria(); ok && val != nil && val.AllOf != nil {
```
Comment (katbyte) [inline azurerm/resource_arm_monitor_metric_alert.go]: same here:

how about we invert this?

```go
if val, ok := v.AsMetricAlertSingleResourceMultipleMetricCriteria(); !ok || val == nil || val.AllOf == nil {
    return result
}
```

File: azurerm/resource_arm_monitor_metric_alert.go [hunk 9]
```diff
@@ -0,0 +1,450 @@
...
+				result = append(result, item)
+			}
+		}
+	}
+	return result
+}
+
+func flattenMonitorMetricAlertAction(v *[]insights.MetricAlertAction) []interface{} {
+	result := make([]interface{}, 0)
+	if v != nil {
```
Comment (katbyte) [inline azurerm/resource_arm_monitor_metric_alert.go]: Could we invert this condition too?

**Review by katbyte — CHANGES_REQUESTED** [review]
Hi @JunyiYi,

Thanks for the new resource, i've given it a quick look and left some minor comments inline. However it appears both the tests and documentation is missing? Once we have those added can do a more in-depth review and get this merged 🙂

**Comment by JunyiYi** [comment]
Thanks @katbyte for the reviewing. Yes it is still in progress (so the title contains `WIP`). Considering this is a similar resource to [`azurerm_monitor_activitylog_alert`](https://github.com/terraform-providers/terraform-provider-azurerm/pull/1989), you might also provide comments in that PR as well. I will combine both to update the code.

File: azurerm/resource_arm_monitor_metric_alert.go (hunk 4 above)
Comment (JunyiYi) [reply azurerm/resource_arm_monitor_metric_alert.go]: I agree. But I'm not sure whether to put it into `utils` or `helpers`?

File: azurerm/resource_arm_monitor_metric_alert.go (hunk 5 above)
Comment (JunyiYi) [reply azurerm/resource_arm_monitor_metric_alert.go]: This comment is out of date.

**Review by katbyte — APPROVED** [review]
@JunyiYi,

LGTM 👍

**Pull request merged**